airflow/decorators/base.py

```diff
@@ -0,0 +1,194 @@
+
+    @staticmethod
+    def validate_python_callable(python_callable):
+        """
+        Validate that python callable can be wrapped by operator.
+        Raises exception if invalid.
+
+        :param python_callable: Python object to be validated
+        :raises: TypeError, AirflowException
+        """
+        if not callable(python_callable):
+            raise TypeError('`python_callable` param must be callable')
+        if 'self' in signature(python_callable).parameters.keys():
+            raise AirflowException('@task does not support methods')
+
+    def execute(self, context: Dict):
+        raise NotImplementedError()
+
+
+T = TypeVar("T", bound=Callable)  # pylint: disable=invalid-name
+
+
+def task_decorator_factory(
+    python_callable: Optional[Callable] = None,
+    multiple_outputs: Optional[bool] = None,
+    decorated_operator_class: BaseDecoratedOperator = None,
+    **kwargs,
+) -> Callable[[T], T]:
+    """
+    A factory that generates a wrapper that raps a function into an Airflow operator.
+    Accepts kwargs for operator kwarg. Can be reused in a single DAG.
+
+    :param python_callable: Function to decorate
+    :type python_callable: Optional[Callable]
+    :param multiple_outputs: if set, function return value will be
+        unrolled to multiple XCom values. List/Tuples will unroll to xcom values
+        with index as key. Dict will unroll to xcom values with keys as XCom keys.
+        Defaults to False.
+    :type multiple_outputs: bool
+    :param decorated_operator_class: The operator that executes the logic needed to run the python function in
+        the correct environment
+    :type decorated_operator_class: BaseDecoratedOperator
+
+    """
+    # try to infer from  type annotation
+    if python_callable and multiple_outputs is None:
+        sig = signature(python_callable).return_annotation
+        ttype = getattr(sig, "__origin__", None)
+
+        multiple_outputs = sig != inspect.Signature.empty and ttype in (dict, Dict)
+
+    def wrapper(f: T):
+        """
+        Python wrapper to generate PythonDecoratedOperator out of simple python functions.
+        Used for Airflow Decorated interface
+        """
+        BaseDecoratedOperator.validate_python_callable(f)
+        kwargs.setdefault('task_id', f.__name__)
+
+        @functools.wraps(f)
+        def factory(*args, **f_kwargs):
+            op = decorated_operator_class(
+                python_callable=f,
+                op_args=args,
+                op_kwargs=f_kwargs,
+                multiple_outputs=multiple_outputs,
+                **kwargs,
+            )
+            if f.__doc__:
+                op.doc_md = f.__doc__
+            return XComArg(op)
+
+        return cast(T, factory)
+
+    if callable(python_callable):
+        return wrapper(python_callable)
+    elif python_callable is not None:
+        raise AirflowException('No args allowed while using @task, use kwargs instead')
+    return wrapper
```

Comment:
How is it possible to get to this final `return wrapper`? What's the scenario where `python_callable` is valid but not a callable?

Reply:
I think this is the difference between `@task` (python_callable will be passed already) and `@task()` -- python_callable won't yet be passed.